### PR TITLE
o/snapstate: validation sets enforcing on update

### PR DESCRIPTION
**Pull request by stolowski**
Enforce validation sets on refreshes.

### CODE REVIEW

File: overlord/snapstate/storehelpers.go
Comment (mardy) [inline overlord/snapstate/storehelpers.go:270]: nitpick: this line could be split

**Review by mardy — APPROVED** [review]
LGTM, just a nitpick you can also ignore :-)

File: overlord/snapstate/storehelpers.go
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:603]: I think this should be a noop too? the user is not really in control of this

File: overlord/snapstate/storehelpers.go
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:456]: same comment about --ignore-validation

File: overlord/snapstate/storehelpers.go
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:337]: I don't think we want a special new error here, we just want to reproduce the behavior if we got back the the current revision from the store which is handled here I think:

https://github.com/snapcore/snapd/blob/master/store/store_action.go#L612

File: overlord/snapstate/refreshhints_test.go
Comment (pedronis) [inline overlord/snapstate/refreshhints_test.go:80]: maybe time to switch to use BaseTest for this?

File: overlord/snapstate/snapstate_remove_test.go
Comment (pedronis) [inline overlord/snapstate/snapstate_remove_test.go:1759]: I missed where this was located, I understand that we started with remove, but maybe this suite if we want to share only  one, should be defined in snapstate_install_test.go ?

**Review by pedronis — COMMENTED** [review]
some comments about errors and the corner of having already the right revision.

also some comments about the test suites

File: overlord/snapstate/storehelpers.go
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:343]: same as install we probably want to ignore it?

File: overlord/snapstate/storehelpers.go
Comment (pedronis) [inline overlord/snapstate/storehelpers.go:606]: we might need to reset cohort in installed in this case?

**Review by pedronis — COMMENTED** [review]
couple comments about cohorts

File: overlord/snapstate/storehelpers.go
Comment (stolowski) [reply overlord/snapstate/storehelpers.go:603]: Indeed, thanks.

**Comment by stolowski** [comment]
Thinking about it more (and looking at the spec) I think I missed some details in the refresh implementation, I think we need to discuss them, I'm marking this blocked till then.

File: overlord/snapstate/refreshhints_test.go
Comment (pedronis) [inline overlord/snapstate/refreshhints_test.go:127]: nitpick: these can also be do with AddCleanup I suppose and drrop the TearDownTest?

**Review by pedronis — COMMENTED** [review]
overall lgtm, wondering if we need to reset the current snaps CohortKey for snaps controlled by validation-sets in more places?

this will need changes once the InstallMany PR has landed

File: overlord/snapstate/storehelpers.go
Comment (stolowski) [reply overlord/snapstate/storehelpers.go:351]: Indeed, done.

**Review by pedronis — APPROVED** [review]
thanks

File: overlord/snapstate/storehelpers.go
Comment (mardy) [inline overlord/snapstate/storehelpers.go:470]: Is it correct to do this even if `requiredValsets` is empty, or should this `if` block be merged together with the next one?

**Review by mardy — APPROVED** [review]
LGTM, with a question.

File: overlord/snapstate/storehelpers.go
Comment (stolowski) [reply overlord/snapstate/storehelpers.go:470]: It is correct, requiredRevision will only be set if there are requiredValsets (in other words, required revision number comes from validation sets that require it). Of course requiredRevision may be unset even if we have requiredValsets here - it depends on whether the validation set assertions have the optional "required: <revision>" attribute for the given snap or not.

**Pull request merged**
